crates/holochain_zome_types/src/query.rs

```diff
@@ -1,26 +1,76 @@
+/// bounded queries MUST be inclusive otherwise the integrity and fork
+/// disambiguation logic is impossible. An exclusive range bound that does not
+/// include the final header tells us nothing about which fork to select
+/// between N forks of equal length that proceed it. With an inclusive hash
+/// bounded range the final header always points unambiguously at the "correct"
+/// fork that the range is over. Start hashes are not needed to provide this
+/// property so ranges can be hash terminted with a length of proceeding
+/// elements to return only. Technically the seq bounded ranges do not imply
+/// any fork disambiguation and so could be a range but for simplicity we left
+/// the API symmetrical in boundedness across all enum variants.
+/// @TODO It may be possible to provide/implement RangeBounds in the case that
+/// a full sequence of elements/headers is provided but it would need to be
+/// handled as inclusive first, to enforce the integrity of the query, then the
+/// exclusiveness achieved by simply removing the final element after the fact.
+#[derive(serde::Serialize, serde::Deserialize, PartialEq, Clone, Debug)]
+pub enum ChainQueryFilterSequenceRange {
+    /// Do NOT apply any range filtering for this query.
+    Unbounded,
+    /// A range over source chain sequence numbers.
+    /// This is ambiguous over forking histories and so should NOT be used in
+    /// validation logic.
+    /// Inclusive start, inclusive end.
+    HeaderSeqRange(u32, u32),
+    /// A range over source chain header hashes.
+    /// This CAN be used in validation logic as forks are disambiguated.
+    /// Inclusive start and end (unlike std::ops::Range).
+    HeaderHashRange(HeaderHash, HeaderHash),
+    /// The terminating header hash and N preceeding elements.
+    /// N = 0 returns only the element with this `HeaderHash`.
+    /// This CAN be used in validation logic as forks are not possible when
+    /// "looking up" towards genesis from some `HeaderHash`.
+    HeaderHashTerminated(HeaderHash, u32),
+}
+
+impl Default for ChainQueryFilterSequenceRange {
+    fn default() -> Self {
+        Self::Unbounded
+    }
+}
+
 /// Query arguments
 #[derive(
     serde::Serialize, serde::Deserialize, SerializedBytes, Default, PartialEq, Clone, Debug,
 )]
 #[non_exhaustive]
 pub struct ChainQueryFilter {
-    /// The range of source chain sequence numbers to match.
-    /// Inclusive start, exclusive end.
-    // TODO: can we generalize this over RangeBounds to allow unbounded ranges?
-    pub sequence_range: Option<std::ops::Range<u32>>,
+    /// Limit the results to a range of elements according to their headers.
+    pub sequence_range: ChainQueryFilterSequenceRange,
     /// Filter by EntryType
     // NB: if this filter is set, you can't verify the results, so don't
     //     use this in validation
     pub entry_type: Option<EntryType>,
+    /// Filter by a list of `EntryHash`.
+    pub entry_hashes: Option<HashSet<EntryHash>>,
```

Comment:
This type gets sent over the network and can now include an unbounded set of hashes?

Reply:
@freesig not sure that the type system is the best place to guard against large network data tbh?